src/chunk.c

```diff
@@ -3278,13 +3190,44 @@ ts_chunk_do_drop_chunks(Hypertable *ht, int64 older_than, int64 newer_than, int3
 	DEBUG_WAITPOINT("drop_chunks_chunks_found");
 
 	if (has_continuous_aggs)
-		ts_chunk_drop_process_invalidations(ht->main_table_relid,
-											older_than,
-											newer_than,
-											chunks,
-											num_chunks);
+	{
+		int i;
 
-	for (; i < num_chunks; i++)
+		/* Exclusively lock all affected chunks and refresh+invalidate the
+		 * dropped regions. Locking prevents further modification of the
+		 * dropped region during this transaction, which allows moving the
+		 * invalidation threshold without having to worry about new
+		 * invalidations while refreshing. */
+		for (i = 0; i < num_chunks; i++)
+		{
+			int64 start = chunks[i].cube->slices[0]->fd.range_start;
+			int64 end = chunks[i].cube->slices[0]->fd.range_end;
+
+			LockRelationOid(chunks[i].table_id, ExclusiveLock);
+
+			Assert(hyperspace_get_open_dimension(ht->space, 0)->fd.id ==
+				   chunks[i].cube->slices[0]->fd.dimension_id);
+
+			/* Refresh all continuous aggregates on the hypertable in the
+			 * region covered by the dropped chunk. Note that we cannot assume
+			 * that all dropped chunks exists in one single contiguous region,
+			 * so we refresh all chunk regions individually rather than the
+			 * region defined by the min and max chunk. An optimization is to
```

Comment:
I changed this from a contiguous range to per-chunk ranges based on @mkindahl's feedback. I think it makes sense to only refresh the regions that change, and not other regions, since the user didn't explicitly specify that this should refresh.

Let's say that you have three contiguous chunks, 1, 2, 3. You manually drop chunk 2 (e.g., DROP TABLE <chunk>). That would leave a gap between chunk 1 and 2. Now, if you later drop everything in a region that includes 1,2,3, then you would refresh also region 2, which would remove all data in the cagg in that region.

I guess I should try to make a test for this case.

Reply:
@erimatnor Your discussion is not applicable here, since this code is part of `ts_chunk_do_drop_chunks`, which takes the drop range as input and then calculates the list of chunks. We know for sure that the requested range is requested by the user. I guess @mkindahl looked only to the diff and hasn't checked where the code belongs.

If you move this code into separate function, which takes a list of chunks, then the refresh per chunk will make sense, but not now.

Reply:
@k-rus the array of chunks here might not cover a contiguous region. I created a test for it.

Reply:
> @k-rus the array of chunks here might not cover a contiguous region. I created a test for it.

I see the only case for this when the chunk doesn't exist in the region, which should not be a problem for refresh. I see a case when the first and last chunks are not aligned with the drop window, but it is not about continuous. Is it the case or are there other cases too?

Reply:
I gave an example of how it would create problems for a refresh. I also created a test case for it. The issue is that a drop invalidates that region, so that a subsequent refresh will bring the region up-to-date with the underlying data. So, if you don't want that to happen, e.g., for the gap in the example, you shouldn't refresh that region; especially not as a hidden side effect of dropping chunks.

Reply:
Here's the diff between doing what you suggest and the current approach of not refreshing the gap:

![image](https://user-images.githubusercontent.com/819732/92638540-b4e63700-f2da-11ea-87b2-5793d0d369d3.png)

Reply:
Sorry, I wasn't able to come up with this case. Basically, it is a drop after earlier drop, which invalidated the region. It is another point to bring for the discussion of removing this refresh before drop functionality. May be the comment can be improved and just mentioned this drop after drop case, and I guess you will add a test to avoid regression.